dbms/src/Flash/Pipeline/Pipeline.cpp

```diff
@@ -32,6 +33,68 @@ FmtBuffer & addPrefix(FmtBuffer & buffer, size_t level)
 {
     return buffer.append(String(level, ' '));
 }
+
+void mapEvents(const Events & inputs, const Events & outputs)
+{
+    assert(!inputs.empty());
+    assert(!outputs.empty());
+    if (inputs.size() == outputs.size())
+    {
+        /**
+         * 1. for fine grained inputs and fine grained outputs
+         *     ```
+         *     FineGrainedPipelineEvent◄────FineGrainedPipelineEvent
+         *     FineGrainedPipelineEvent◄────FineGrainedPipelineEvent
+         *     FineGrainedPipelineEvent◄────FineGrainedPipelineEvent
+         *     FineGrainedPipelineEvent◄────FineGrainedPipelineEvent
+         *     ```
+         * 2. for non fine grained inputs and non fine grained outputs
+         *     ```
+         *     PlainPipelineEvent◄────PlainPipelineEvent
+         *     ```
+         * 3. for non fine grained inputs and fine grained outputs
+         *     This is not possible, if fine-grained is enabled in outputs, then inputs must also be enabled.
+         *     Checked in `doToEvents`.
+         * 4. for fine grained inputs and non fine grained outputs
+         *     ```
+         *     PlainPipelineEvent◄────FineGrainedPipelineEvent
+         *     ```
+         */
+        size_t partition_num = inputs.size();
+        for (size_t index = 0; index < partition_num; ++index)
+            outputs[index]->addInput(inputs[index]);
+    }
+    else
+    {
+        /**
+         * 1. for fine grained inputs and fine grained outputs
+         *     If the number of partitions does not match, it is safer to use full mapping.
+         *     ```
+         *     FineGrainedPipelineEvent◄──┐ ┌──FineGrainedPipelineEvent
+         *     FineGrainedPipelineEvent◄──┼─┼──FineGrainedPipelineEvent
+         *     FineGrainedPipelineEvent◄──┤ └──FineGrainedPipelineEvent
+         *     FineGrainedPipelineEvent◄──┘
+         *     ```
```

Comment:
Then the next pipeline's FineGrainedPipelineEvent must relay on all FineGrainedPipelineEvents before this pipeline?

Reply:
yes

Reply:
Maybe add some log here to identify different scene.

Reply:
Adding logs here is meaningless now, because there is only one fine grained pipeline at most...